core/src/main/scala/kafka/server/DynamicBrokerConfig.scala

```diff
@@ -449,3 +451,63 @@ class DynamicLogConfig(logManager: LogManager) extends Reconfigurable with Loggi
     }
   }
 }
+
+object DynamicThreadPool {
+  val ReconfigurableConfigs = Set(
+    KafkaConfig.NumIoThreadsProp,
+    KafkaConfig.NumNetworkThreadsProp,
+    KafkaConfig.NumReplicaFetchersProp,
+    KafkaConfig.NumRecoveryThreadsPerDataDirProp,
+    KafkaConfig.BackgroundThreadsProp)
+}
+
+class DynamicThreadPool(server: KafkaServer) extends Reconfigurable {
+
+  override def configure(configs: util.Map[String, _]): Unit = {}
+
+  override def reconfigurableConfigs(): util.Set[String] = {
+    DynamicThreadPool.ReconfigurableConfigs.asJava
+  }
+
+  override def validateReconfiguration(configs: util.Map[String, _]): Boolean = {
+    configs.asScala.filterKeys(DynamicThreadPool.ReconfigurableConfigs.contains).forall { case (k, v) =>
+      val newValue = v.asInstanceOf[Int]
+      val oldValue = currentValue(k)
+      if (newValue != oldValue) {
+        val errorMsg = s"Dynamic thread count update validation failed for $k=$v"
+        if (newValue <= 0)
+          throw new ConfigException(s"$errorMsg, value should be at least 1")
+        if (newValue < oldValue / 2)
```

Comment:
Just so I understand, is the purpose of this check to ensure that configs are only changed "gracefully."

Reply:
Yes, since threads can have a big impact, I thought it was better to limit the values. The updates will be more graceful and it will avoid a wrong config bringing down brokers.